### ISSUE TITLE
Memory leak in BrowserSubprocess.exe when cefSharp.bindObjectAsync is used

### ISSUE THREAD

**Issue by ekalchev** [cefsharp/CefSharp#2825]
- **What version of the product are you using?**

I found it in 67 and upgraded to 73 and it is still present.

- **What architecture x86 or x64?**

x86

- **On what operating system?**
    - Win7, Win 8, Win10, etc?

Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**

WinForms

- **Does this problem also occur in the `CEF` Sample Application from http://opensource.spotify.com/cefbuilds/index.html?**

No

So I have application that loads a custom scheme/ domain (http://mail-viewer) URL where I am showing html from emails.
For different emails I am re-using the domain above and a new SchemeHandler is registered each time that provide different HTML.

For each email I am calling this line

cefSharp.bindObjectAsync("logger");

I noticed that memory consumption of BrowserSubprocess.exe is jumping with 20mb - 30mb each time I am loading new email and forcing GC from Dev tools is not releasing the memory. The memory raises to around 1.3Gb and then the process crash.

After some investigation I found that commenting this line (I dont have other bindObjectAsync calls)

cefSharp.bindObjectAsync("logger");

 in my javascript file is resolving the issue. It seems that call is causing a memory leak. I tried to reproduce it in the examples project but I couldn't. It seems it is something related to bindObjectAsync

**Comment by ekalchev** [comment]
<a href="https://ibb.co/GxNgK9D"><img src="https://i.ibb.co/5MwqNk3/Annotation-2019-07-02-181431.png" alt="Annotation-2019-07-02-181431" border="0"></a>

**Comment by ekalchev** [comment]
I noticed that when have memory profiler attached to the BrowserProcess.exe sometimes the memory is released when I am making a memory snapshot.

**Comment by amaitland** [comment]
As per the issue template template please fork https://github.com/cefsharp/CefSharp.MinimalExample and provide an example that reproduces the problem. 

If taking a memory snapshot sometimes releases the memory then this is more likely a CEF bug.

**Comment by ekalchev** [comment]
I was able to reproduce it with CefSharp.MinimalExample
https://github.com/ekalchev/CefSharp.MinimalExample

What I did is to create a custom scheme handler to return static HTML page. Inside the HTML page there is this code (other is just copy from Wikipedia page)

<script>
cefSharp.bindObjectAsync("logger");
</script>

The app is reloading the page over and over to simulate mail client which open emails. You can see the process memory is growing up. At my computer it crashed at around 3gb.

If you comment the bindObjectAsync call the memory stop growing and appear to be working without a leak (or the leak is very small and not noticeable)

**Comment by amaitland** [comment]
> Inside the HTML page there is this code (other is just copy from Wikipedia page)

Is the extra html and images necessary to reproduce the issue? If not then please reduce your example down to the very minimum required.

If you need further clarification then please read https://stackoverflow.com/help/minimal-reproducible-example

**Comment by ekalchev** [comment]
You need to load page that takes some memory to notice the memory leak, because memory usage builds up on every consecutive load. With empty page the leak is not noticeable. It doesn't matter if it is some heavy js library or alot of static html. That is why I copied a page from wikipedia.

**Comment by amaitland** [comment]
> With empty page the leak is not noticeable. It doesn't matter if it is some heavy js library or alot of static html. That is why I copied a page from wikipedia.

Again that would indicate a bug in `CEF/Chromium`. The exact same code is run from a `CefSharp` point of view, the same amount of memory should be leaked every time.

Your code contains all sorts of command line args and other pieces that I have no idea if are relevant, it would be very helpful if you could reduce the example down to the bare minimum that's required to reproduce the issue, then squash all your commits together.

Also I downloaded your project and searched for `cefSharp.bindObjectAsync` solution wide and found no reference.

**Comment by ekalchev** [comment]
I did what you asked. 

cefSharp.bindObjectAsync was inside html file which wasn't part of the csproj file. I added it so you can find it.

**Comment by amaitland** [comment]
Great,  thank you. Makes a huge difference.

I will have a look later today.

**Comment by amaitland** [comment]
Confirming that I can reproduce the issue.

I'm not seeing any noticeable leak when running the `Javascript Binding Stress Test`. More investigation required.

**Comment by amaitland** [comment]
I made some changes that will be in the upcoming `75` release, specifically when a `Bound Object` is available in the cache then it will resolve directly without creating a custom promise, this in reality should resolve any real world memory leaks. See https://github.com/cefsharp/CefSharp/commit/31fdcd459beb86a7903ab3bf97ce7fa0899e9930#diff-ddf74c619f8a763151f502a2db7480a5

Calling `Load/Reload` in a tight loop unfortunately doesn't really give the `V8 GC` time to run properly, it's starved effectively. There's other instances of this happening in relation to `CEF` when executing `JS`. There isn't anything I can do about that from a `CefSharp` point of view, the `C++ Object Destructors`  simply aren't called. The change I made above avoids creating an object, so that's the best workaround I can come up with.

An additional change to make sure the `JavascriptAsyncMethodCallback` is deleted was made in commit https://github.com/cefsharp/CefSharp/commit/00d9b59b5529bc70abf5329b625d7f5de92405a0#diff-8dabe4228c63cbb1fba228450b3b7609

Loading a page with the following call in a tight loop using https://github.com/cefsharp/CefSharp/commit/bbef0f3fa6ece1a4fc5e2a7403d32e052f5def94 gives me semi stable memory usage, gets to about `400mb` then frees some of it. If you were to call `bindObjectAsync` for an object that actually gets bound (and sufficiently has time to complete the binding process before the `V8 Context` is released) then there shouldn't be any memory leak.

```js
cefSharp.bindObjectAsync("something");
```

By all means update your example to the https://github.com/cefsharp/CefSharp/releases/tag/v75.1.140-pre01 and see what you get. 

This issue will likely be closed as it's unlikely we can make any further improvements.
